diff --git a/src/util/lp/lp_solver.cpp b/src/util/lp/lp_solver.cpp
index ea88dd14c..2a1b89113 100644
--- a/src/util/lp/lp_solver.cpp
+++ b/src/util/lp/lp_solver.cpp
@@ -44,6 +44,8 @@ template void lp::lp_solver<lp::mpq, lp::mpq>::flip_costs();
 template lp::mpq lp::lp_solver<lp::mpq, lp::mpq>::get_column_cost_value(unsigned int, lp::column_info<lp::mpq>*) const;
 template int lp::lp_solver<lp::mpq, lp::mpq>::get_column_index_by_name(std::string) const;
 template lp::mpq lp::lp_solver<lp::mpq, lp::mpq>::get_column_value_by_name(std::string) const;
+template std::string lp::lp_solver<double, double>::get_column_name(unsigned int) const;
+template std::string lp::lp_solver<lp::mpq, lp::mpq>::get_column_name(unsigned int) const;
 template lp::mpq lp::lp_solver<lp::mpq, lp::mpq>::get_column_value_with_core_solver(unsigned int, lp::lp_core_solver_base<lp::mpq, lp::mpq>*) const;
 template lp::column_info<lp::mpq>* lp::lp_solver<lp::mpq, lp::mpq>::get_or_create_column_info(unsigned int);
 template void lp::lp_solver<lp::mpq, lp::mpq>::give_symbolic_name_to_column(std::string, unsigned int);
